backend/tracim_backend/lib/core/event.py

```diff
@@ -299,19 +315,34 @@ def _create_role_event(
             _MEMBER_FIELD: self._workspace_user_role_schema.dump(role_in_context).data,
         }
         event = Event(entity_type=EntityType.WORKSPACE_MEMBER, operation=operation, fields=fields)
+        self._add_event(event, context)
+
+    def _add_event(self, event: Event, context: TracimContext) -> None:
         context.dbsession.add(event)
+        context.pending_events.append(event)
 
-    def _publish_event(self, session: TracimSession, instance: object) -> None:
-        if not isinstance(instance, Event):
-            return
-        event = typing.cast(Event, instance)
+    def _publish_events(self, context: TracimContext) -> None:
         if self._config.JOBS__PROCESSING_MODE == self._config.CST.ASYNC:
             message_builder = AsyncLiveMessageBuilder(
-                context=session.context
+                context=context
             )  # type: BaseLiveMessageBuilder
         else:
-            message_builder = SyncLiveMessageBuilder(context=session.context)
-        message_builder.publish_messages_for_event(event.event_id)
+            message_builder = SyncLiveMessageBuilder(context=context)
+
+        # We only publish events that have an event_id from the DB.
+        # we can have `new` events here as we add events in the session
```

Comment:
Not sure to understand why we can have unpublishable event, do you have an exemple ?

Reply:
The event creation process is done like this (example for user):

1. user_api.update(user)
2. (during session `after_flush`  number 1):
   - on_user_modified
   - create_user_event, add it to session
   - `_publish_events` is also called as it is a `post_flush` operation when processing mode is `sync`. At this point the `user.modified` event does not have its id set (it has not been flushed yet, only added to the session)
3. (during session `after_flush` number 2, eg. just before transaction.commit):
  - `_publish_events` is called again, `user.modified` pending event has its id set so can be published.

What's a bit unfortunate is that we cannot do any SQL with the session's in `after_commit`: at this point we are sure that all pending events have an id, but we cannot use the session to do SQL anymore.

Reply:
ok.